### PR TITLE
fix(rpc): Fix status `CatchingUp` field updating

### LINKED ISSUE #935: `/status` endpoints `sync_info.catching_up` field always shows false
`/status` endpoints `sync_info.catching_up` field always shows false, even when a full node is behind

### PR DESCRIPTION
**Pull request by zale144**
# PR Standards

## Opening a pull request should be able to meet the following requirements

--

PR naming convention: https://hackmd.io/@nZpxHZ0CT7O5ngTp0TP9mg/HJP_jrm7A

---

Close #935

Basically, the idea is to have the CatchingUp value be derived as such:

`CatchingUp = targetHeight > latestHeight`

My guess is that there might need to be some tolerance there, in case a new settlement batch is submitted, and at the same moment the `/status` is queried.

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

File: rpc/client/client.go [hunk 1]
```diff
@@ -751,13 +748,13 @@ func (c *Client) Status(ctx context.Context) (*ctypes.ResultStatus, error) {
 			LatestAppHash:     latestAppHash[:],
 			LatestBlockHeight: int64(latestHeight),
 			LatestBlockTime:   time.Unix(0, int64(latestBlockTimeNano)),
+			CatchingUp:        c.node.BlockManager.TargetHeight.Load() > latestHeight,
```
Comment (mtsitrin) [inline rpc/client/client.go:751]: u need to set `TargetHeight` also in  `syncBlockManager` , currnetly u made it event based only

File: block/manager.go [hunk 2]
```diff
@@ -64,6 +64,7 @@ type Manager struct {
 	// It is ALSO used by the producer, because the producer needs to check if it can prune blocks and it wont'
 	// prune anything that might be submitted in the future. Therefore, it must be atomic.
 	LastSubmittedHeight atomic.Uint64
+	TargetHeight        atomic.Uint64
```
Comment (mtsitrin) [inline block/manager.go:67]: I think it's redundent. both `LastSubmittedHeight` and `TargetHeight` are actually `LatestHeightOnSL`

**Review by mtsitrin — CHANGES_REQUESTED** [review]

File: block/gossip.go [hunk 3]
```diff
@@ -22,6 +23,10 @@ func (m *Manager) onNewGossipedBlock(event pubsub.Message) {
 		return
 	}
 
+	if block.Header.Height > m.TargetHeight.Load() {
+		m.TargetHeight.Store(block.Header.Height)
+	}
```
Comment (danwt) [inline block/gossip.go:28]: this is not concurrency safe you need to use compareAndSwap
and please use stdlib max

File: block/manager.go [hunk 4]
```diff
@@ -55,9 +55,10 @@ type Manager struct {
 	*/
 	// The last height which was submitted to both sublayers, that we know of. When we produce new batches, we will
 	// start at this height + 1.
-	// It is ALSO used by the producer, because the producer needs to check if it can prune blocks and it wont'
+	// It is ALSO used by the producer, because the producer needs to check if it can prune blocks and it won't
 	// prune anything that might be submitted in the future. Therefore, it must be atomic.
 	LastSubmittedHeight atomic.Uint64
+	TargetHeight        atomic.Uint64
```
Comment (danwt) [inline block/manager.go:61]: It's under the submission section but it's used in the retriever loop

Needs a docstring too, I don't understand how it relates (or doesn't) to targetSyncHeight

**Review by danwt — CHANGES_REQUESTED** [review]

File: rpc/client/client.go (hunk 1 above)
Comment (danwt) [reply rpc/client/client.go:750]: this field needs a docstring as well IMO
the other ones are obvious and fine

**Review by danwt — CHANGES_REQUESTED** [review]
I'm still not sure how this relates to the targetSyncHeight field and why we can just have one field

File: rpc/client/client.go (hunk 1 above)
Comment (zale144) [reply rpc/client/client.go]: fix pushed

File: rpc/client/client.go [hunk 5]
```diff
@@ -750,13 +747,14 @@ func (c *Client) Status(ctx context.Context) (*ctypes.ResultStatus, error) {
 			LatestAppHash:     latestAppHash[:],
 			LatestBlockHeight: int64(latestHeight),
 			LatestBlockTime:   time.Unix(0, int64(latestBlockTimeNano)),
+			// CatchingUp is true if the node is not at the latest height received from p2p or da.
```
Comment (danwt) [inline rpc/client/client.go:750]: shuld be on the field on the struct, not where the value is set!

Comment (zale144) [reply rpc/client/client.go:750]: should I fork Tendermint then?

Comment (danwt) [reply rpc/client/client.go:750]: ah sorry didn't realise it was tendermint
all good

File: block/manager.go [hunk 6]
```diff
@@ -225,3 +228,12 @@ func (m *Manager) syncBlockManager() error {
 	m.logger.Info("Synced.", "current height", m.State.Height(), "last submitted height", m.LastSubmittedHeight.Load())
 	return nil
 }
+
+func (m *Manager) updateTargetHeight(h uint64) {
+	for {
+		currentHeight := m.TargetHeight.Load()
+		if m.TargetHeight.CompareAndSwap(currentHeight, max(currentHeight, h)) {
```
Comment (mtsitrin) [inline block/manager.go:235]: if `max(currentHeight, h)` is `currentHeight`
u'll call `m.TargetHeight.CompareAndSwap(currentHeight,currentHeight)`
won't it return `false` and be stuck forever?

**Review by mtsitrin — APPROVED** [review]

**Pull request merged**